sos/deviceauth.py

```diff
@@ -0,0 +1,294 @@
+# Copyright (C) 2023 Red Hat, Inc., Jose Castillo <jcastillo@redhat.com>
+
+# This file is part of the sos project: https://github.com/sosreport/sos
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# version 2 of the GNU General Public License.
+#
+# See the LICENSE file in the source distribution for further information.
+
+import logging
+import requests
+import time
+from datetime import datetime, timedelta
+import os
+from sos.keyring import Key, KeyNotFoundError
+from sos.utilities import is_executable
+
+DEVICE_AUTH_CLIENT_ID = "sos-tools"
+GRANT_TYPE_DEVICE_CODE = "urn:ietf:params:oauth:grant-type:device_code"
+DATETIME_FORMAT = "%Y-%m-%dT%H:%M:%SZ"
+
+logger = logging.getLogger("sos")
+
+
+def try_read_refresh_token():
+    """Try to read locally stored refresh token
+
+    Returns:
+        str: Returns ODIC refresh token if found otherwise None
+    """
+    try:
+        AuthKey = Key()
+        key_refresh_token = AuthKey.search('sos-tools_refresh_token')
+        key_username = AuthKey.search('sos-tools-user')
+    except KeyNotFoundError:
+        logger.info("Refresh token does not exist in keyring or is expired.")
+        return None
+    except Exception as e:
+        logger.error("Error encoutered while accessing keyring. {}".format(e))
+        return None
+    if key_username.data.decode() != os.getlogin():
+        return None
```

Comment:
I will need to revisit this because it seems that produces more problems than benefits, specially since we set up session tokens, so it will never be another user trying to read the keyring.